### PR TITLE
add config.services.xserver.windowManager.sway

### PR DESCRIPTION
**Pull request by tg-x**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by tg-x** [comment]
needs wayland-compatible display manager such as SDDM or GDM (tested with SDDM)

**Comment by Ericson2314** [comment]
Is this x+wayland? Either way its good progress!

**Comment by tg-x** [comment]
yes, the display manager still runs on X, this just adds the option to select sway, that would run in a wayland session

**Comment by groxxda** [comment]
Thanks for the work. Obviously this should not be part of `services.xserver`, but fixing this is a whole other issue. So it's probably ok for now.

Another thing:
At least GDM decides which session (x or wayland) is needed by looking at the **path** of the `.desktop` file. If I remember correctly, support for `GDM-Session-Type` (or similar) was dropped again.

I would really appreciate it, if you tested this with gdm 3.22 from #19081 which itself uses wayland instead of X and see if it works. I might give it a shot myself when I find time.

File: nixos/modules/services/x11/window-managers/sway.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      '';
+    }];
+    environment.systemPackages = [ pkg ];
+  };
+
+in
+
+{
+  options.services.xserver.windowManager = {
+    sway = swayOption "sway";
```
Comment (ericsagnes) [inline nixos/modules/services/x11/window-managers/sway.nix]: Why using a function to generate the options?

File: nixos/modules/services/x11/window-managers/sway.nix [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  };
+
+  config = mkMerge [
+    (mkIf wmCfg.sway.enable (swayConfig "sway" pkgs.sway wmCfg.sway))
```
Comment (ericsagnes) [inline nixos/modules/services/x11/window-managers/sway.nix]: Why using a `mkMerge` and a function to generate the config?

File: nixos/modules/services/x11/window-managers/sway.nix (hunk 1 above)
Comment (tg-x) [inline nixos/modules/services/x11/window-managers/sway.nix]: this is based on i3.nix, sure it does not need the function or mkMerge anymore, could be simplified..

Comment (ericsagnes) [inline nixos/modules/services/x11/window-managers/sway.nix]: Please do so, that will improve the readability and maintainability of the module.
For the record, i3 module was simplified in #20385.

**Comment by vikstrous** [comment]
I tested this out. I was using slim. Sway started and it even loaded my i3 config and ran i3status without me creating a separate sway config. However, whenever I tried to create a terminal, the process started but it was not visible on the screen. Looking at the process tree in htop I could see all of my terminals and they were all under pid 1. Everything looked normal, I just couldn't see them.

**Comment by matthiasbeyer** [comment]
This is awesome progress!

Note that in https://github.com/NixOS/nixpkgs/pull/21808 sway will be updated, which might result in unexpected things after the merge.

I would vote for an `options.services.graphics.xserver` option, keeping `options.services.xserver` to be backwards compatible.
But sure, this is another issue and should be discussed elsewhere.

**Comment by Mic92** [comment]
Can you add a setcap wrapper as described in the [README.md](https://github.com/SirCmpwn/sway#compiling-from-source) using our new [setcap wrapper](https://github.com/NixOS/nixpkgs/pull/16654)?

**Comment by colemickens** [comment]
This can be closed. It's not longer relevant or up-to-date.

**Comment by matthew-piziak** [comment]
@colemickens So Sway will not be supported in NixOS?

**Comment by colemickens** [comment]
@matthew-piziak oh no, I'm sorry I didn't explain more. There is already a Sway package and module in nixpkgs now!

**Comment by bobvanderlinden** [comment]
Reference: https://github.com/NixOS/nixpkgs/blob/release-18.09/nixos/modules/programs/sway.nix

**Comment by matthew-piziak** [comment]
Weird, if I add `programs.sway.enable = true;` to my `configuration.nix` and also put `"sway"` in my `extraGroups`, then running `sway` gives me `Couldn't exec sway-setcap: Permission denied`.

Getting a permission error is weird, given that I'm part of the group.

If I run `sway` as root, I get `03/19/19 13:42:51 - [main.c:48] [wlc] Failed to add socket to wayland display`.

**Comment by colemickens** [comment]
You're not up to date if you're getting errors about setcap and wlc.

See this commit: https://github.com/NixOS/nixpkgs-channels/commit/45004c6f6330b1ff6f3d6c3a0ea8019f6c18a930 It's in nixos-unstable and I know the change was ported into the 19.03 channel as well.

**Comment by matthew-piziak** [comment]
Ah, thanks. Unfortunately NVIDIA drivers do not build on 19.03, and sway does not build on 18.09. NVIDIA is more important to me. I haven't figured out how to import program configuration from specific channels (like sway from unstable, for example). I only know how to do it for packages, not programs. I don't suppose you know how to do that?

**Comment by colemickens** [comment]
1. Sway 1.0 explicitly will refuse to start with nvidia drivers, anyway. You'd need to use nouvaeu.
2. The NUR supposedly has examples of importing modules but I've never gotten it to work. I actually wanted to do this as part of [nixpkgs-wayland](https://github.com/colemickens/nixpkgs-wayland).

**Comment by matthew-piziak** [comment]
Oh, and there's no way to use one set of drivers for sway and one for everything else right? Probably a silly question but I really don't know. Prooobably not going to give up the graphics card performance in exchange for this window manager.

Thanks for answering my questions!

**Comment by bobvanderlinden** [comment]
Which features are in this PR that aren't on master? If there aren't any, please close this PR.

**Comment by moaxcp** [comment]
> Which features are in this PR that aren't on master? If there aren't any, please close this PR. (making an attempt to clean up older PRs)

I don't see any of these features in master.

**Comment by colemickens** [comment]
Aren't they on master under `programs.sway` ? https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/sway.nix

**Comment by moaxcp** [comment]
> Aren't they on master under `programs.sway` ? https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/sway.nix

Yes they are! It makes sense to close then.

**Pull request closed without merging**